src/test/java/at/ac/tuwien/kr/alpha/solver/SolverTests.java

```diff
@@ -558,6 +567,58 @@ public void groundAtomInRule() throws IOException {
 		);
 	}
 
+	@Test
+	public void simpleChoiceRule() throws IOException {
+		assertAnswerSetsWithBase(
+			"{ a; b; c} :- d." +
+				"d.",
+
+			"d",
+			"",
+			"a",
+			"a, b",
+			"a, c",
+			"a, b, c",
+			"b",
+			"b, c",
+			"c"
+		);
+	}
+
+	@Test
+	public void conditionalChoiceRule() throws IOException {
+		assertAnswerSetsWithBase(
+			"dom(1..3)." +
+				"{ p(X): not q(X); r(Y): p(Y)} :- dom(X), q(Y)." +
+				"q(2).",
+
+			"dom(1)," +
+				"dom(2)," +
+				"dom(3)," +
+				"q(2)",
+
+			"p(1)," +
+				"p(3)",
+
+			"",
+
+			"p(3)",
+
+			"p(1)"
+		);
+	}
+
+	@Test
+	public void doubleChoiceRule() throws IOException {
+		Program program = parser.parse("{ a }. { a }.");
+		Solver solver = getInstance(new NaiveGrounder(program));
+		int countAnswerSets = 0;
+		while (solver.spliterator().tryAdvance(d -> { })) {
+			countAnswerSets++;
+		}
+		assertEquals(2, countAnswerSets);
```

Comment:
There's no need to go so low-level here.

```java
assertEquals(2, solve("{ a }. { a }.").size());
```

Reply:
I had my reasons for that as follows: choice rules introduce a rewriting with hidden predicates (basically `_na :- not a.  a :- not _na.` twice) and the number of answer-sets before projection of hidden predicates may be different than after projection.

The high-level collection of answer-sets is based on the projected answer-sets collected in a `Set<AnswerSet>`. So the solver might find two distinct answer-sets `{ _na }`and `{ _na' }` but after projection they both become `{ }`. The `Set<AnswerSet>` then swallows one of the answer-sets actually found be the solver. By goind low-level as in the test, it should be possible to observe them both, however.

This is the reason, why I think the high-level counting may not work as intended by the test. I'm happy to learn of a better method for checking the actually found answer-sets, though.

Reply:
Oh, yeah. In this case you could use `solver.collectList()` which should give you a list containing two elements, both of which are equal to the empty answer set.

```
List<AnswerSet> expected = Arrays.asList(BasicAnswerSet.EMPTY, BasicAnswerSet.EMPTY);
List<AnswerSet> actual = solver.collectList();
assertEquals(expected, actual);
```

Reply:
The answer sets for this program indeed are `{}, { a }` but the test shall ensure that no further answer sets (which differ only on hidden atoms) are reported. So if there were two "same" answer sets, the `collectList()` should see them.